compiler/src/org.graalvm.compiler.replacements/src/org/graalvm/compiler/replacements/StandardGraphBuilderPlugins.java

```diff
@@ -1171,24 +1140,56 @@ public boolean apply(GraphBuilderContext b, ResolvedJavaMethod targetMethod, Rec
             // Opaque mode does not directly impose any ordering constraints with respect to other
             // variables beyond Plain mode.
-            if (accessKind == AccessKind.OPAQUE && StampTool.isPointerAlwaysNull(object)) {
+            if (memoryBarriersKind == MemoryBarriersKind.OPAQUE && StampTool.isPointerAlwaysNull(object)) {
                 // OFF_HEAP_LOCATION accesses are not floatable => no membars needed for opaque.
                 return apply(b, targetMethod, unsafe, offset, value);
             }
             // Emits a null-check for the otherwise unused receiver
             unsafe.get();
-            boolean isVolatile = accessKind == AccessKind.VOLATILE;
-            boolean emitBarriers = accessKind.emitBarriers && !isVolatile;
+            boolean isVolatile = memoryBarriersKind == MemoryBarriersKind.VOLATILE;
+            boolean emitBarriers = memoryBarriersKind.emitBarriers && !isVolatile;
             if (emitBarriers) {
-                b.add(new MembarNode(accessKind.preWriteBarriers));
+                b.add(new MembarNode(memoryBarriersKind.preWriteBarriers));
             }
             ValueNode maskedValue = b.maskSubWordValue(value, unsafeAccessKind);
             // Raw accesses can be turned into floatable field accesses, the membars preserve the
             // access mode. In the case of opaque access, and only for opaque, the location of the
             // wrapping membars can be refined to the field location.
             createUnsafeAccess(object, b, (obj, loc) -> new RawStoreNode(obj, offset, maskedValue, unsafeAccessKind, loc, true, isVolatile));
             if (emitBarriers) {
-                b.add(new MembarNode(accessKind.postWriteBarriers));
+                b.add(new MembarNode(memoryBarriersKind.postWriteBarriers));
+            }
+            return true;
+        }
+    }
+
+    public static class UnsafeCompareAndSwapPlugin extends UnsafeAccessPlugin {
+        private final MemoryBarriersKind memoryBarriersKind;
+        private final JavaKind accessKind;
+        private final boolean isLogic;
+
+        public UnsafeCompareAndSwapPlugin(JavaKind returnKind, JavaKind accessKind, MemoryBarriersKind memoryBarriersKind, boolean isLogic, boolean explicitUnsafeNullChecks) {
+            super(returnKind, explicitUnsafeNullChecks);
+            this.memoryBarriersKind = memoryBarriersKind;
+            this.accessKind = accessKind;
+            this.isLogic = isLogic;
+        }
+
+        @Override
+        public boolean apply(GraphBuilderContext b, ResolvedJavaMethod targetMethod, Receiver unsafe, ValueNode object, ValueNode offset, ValueNode expected, ValueNode newValue) {
+            // Emits a null-check for the otherwise unused receiver
+            unsafe.get();
+            if (memoryBarriersKind == MemoryBarriersKind.RELEASE) {
```

Comment:
Enforcing the orderings can be done directly within the CompareAndSwapNodes so that these memory barriers are not needed. See my other general comment